src/Storage/OWSIncomingMessageFinder.m

```diff
@@ -0,0 +1,156 @@
+
+    return [self initWithDatabase:[TSStorageManager sharedManager].database];
+}
+
+- (instancetype)initWithDatabase:(YapDatabase *)database
+{
+    self = [super init];
+    if (!self) {
+        return self;
+    }
+
+    _database = database;
+
+    return self;
+}
+
+#pragma mark - properties
+
+- (YapDatabaseConnection *)dbConnection
+{
+    @synchronized (self) {
+        if (!_dbConnection) {
+            _dbConnection = self.database.newConnection;
+        }
+    }
+    return _dbConnection;
+}
+
+#pragma mark - YAP integration
+
+- (YapDatabaseSecondaryIndex *)indexExtension
+{
+    YapDatabaseSecondaryIndexSetup *setup = [YapDatabaseSecondaryIndexSetup new];
+
+    [setup addColumn:OWSIncomingMessageFinderColumnTimestamp withType:YapDatabaseSecondaryIndexTypeInteger];
+    [setup addColumn:OWSIncomingMessageFinderColumnSourceId withType:YapDatabaseSecondaryIndexTypeText];
+    [setup addColumn:OWSIncomingMessageFinderColumnSourceDeviceId withType:YapDatabaseSecondaryIndexTypeInteger];
+
+    YapDatabaseSecondaryIndexWithObjectBlock block = ^(YapDatabaseReadTransaction *transaction,
+        NSMutableDictionary *dict,
+        NSString *collection,
+        NSString *key,
+        id object) {
+        if ([object isKindOfClass:[TSIncomingMessage class]]) {
+            TSIncomingMessage *incomingMessage = (TSIncomingMessage *)object;
+
+            [dict setObject:@(incomingMessage.timestamp) forKey:OWSIncomingMessageFinderColumnTimestamp];
+            [dict setObject:incomingMessage.authorId forKey:OWSIncomingMessageFinderColumnSourceId];
+            [dict setObject:@(incomingMessage.sourceDeviceId) forKey:OWSIncomingMessageFinderColumnSourceDeviceId];
+        }
+    };
+
+    YapDatabaseSecondaryIndexHandler *handler = [YapDatabaseSecondaryIndexHandler withObjectBlock:block];
+
+    return [[YapDatabaseSecondaryIndex alloc] initWithSetup:setup handler:handler];
+}
+
+- (void)asyncRegisterExtension
+{
+    DDLogInfo(@"%@ registering async.", self.tag);
+    [self.database registerExtension:self.indexExtension withName:OWSIncomingMessageFinderExtensionName];
+}
+
+- (void)registerExtension
+{
+    OWSAssert(NO);
+    DDLogError(@"%@ registering SYNC. We should prefer async when possible.", self.tag);
+    [self.database registerExtension:self.indexExtension withName:OWSIncomingMessageFinderExtensionName];
+}
+
+#pragma mark - instance methods
+
+- (BOOL)existsMessageWithTimestamp:(uint64_t)timestamp
+                          sourceId:(NSString *)sourceId
+                    sourceDeviceId:(uint32_t)sourceDeviceId
+{
+    if (![self.database registeredExtension:OWSIncomingMessageFinderExtensionName]) {
+        DDLogError(@"%@ in %s but extension is not registered", self.tag, __PRETTY_FUNCTION__);
+        OWSAssert(NO);
+
+        // we should be initializing this at startup rather than have an unexpectedly slow lazy setup at random.
+        [self registerExtension];
```

Comment:
Is this to handle the race condition where we receive an incoming message before `asyncRegisterExtension` is called?  `asyncRegisterExtension` will always be called - is there any harm to registering twice?  Or is that why `asyncRegisterExtension` is actually synchronous?  If so, can't we remove this check? etc.